### DISCUSSION TITLE
Equal Cost MultiPath (ECMP) implementation - INET4.2.1

### DISCUSSION THREAD

**Discussion by lakshmikrish-97** [inet-framework/inet#643] [Ideas]
Hi all!

I am new to OMNeT++ and INET framework. I am looking to run some data center network simulations and was wondering if there exists any implementation of ECMP module for routing flows in a data center network setup?

**Comment by Sougol** [comment]
Following on this thread! 
Besides, I was wondering if any kind of multipathing has been added to the Inet framework?

**Reply by heychr** [reply]
hello, can you please let me know if you got an answer? is there any multipath routing added to inet or omnet?

**Comment by lgc0208** [comment]
Following too!

**Comment by heychr** [comment]
hello, can you please let me know if you got an answer? is there any multipath routing added to inet or omnet?

**Comment by avarga** [comment]
A Google search gave me this repo:

https://gitraap.i3a.info/jfclemente/ecmp. "ECMP INET 4.4"

**Reply by heychr** [reply]
hello, i am having trouble to build this framework, does it work for you?

**Comment by avarga** [comment]
We recently added ECMP to opp_env's database, so if you are on Linux or macOS (or use WSL on Windows), you can install it along with with OMNeT++ with a single command:

    opp_env install -w ecmp-workspace --init ecmp_allinone

Then you can go grab some coffee or tea and enjoy it while opp_env downloads the packages, unpacks them, and compiles the projects. Then change into the new `ecmp-workspace/` directory, and enter

    opp_env shell

to get a shell prompt where you can work with the ECMP model.

This installs from: https://github.com/inet-framework/inet-clos-ecmp/. The following commit contains all the contributions the ECMP model makes to INET-4.4.1: https://github.com/inet-framework/inet-clos-ecmp/commit/2f5644ce0b8061b3759a68164d01225a845a65a6

You can grab opp_env from here: https://github.com/omnetpp/opp_env#opp_env-automated-installation-of-omnet-simulation-frameworks.

**Reply by JayHarden12** [reply]
Hi M. Andras, does the framework support ECMP on L3; I'm trying to work on my BSc. research that compares RIP, OSPF and ECMP OSPF I'm using OMNet++ to carry out this research.

Thanks,
Jay

**Reply by rhornig** [reply]
We are not the maintainer or author of the project. We made it available for easier installation in opp_env, but we do not have any insight about the implementation or suitability for any purpose. You should try to install it and try to figure out whether that's suitable for you (or contact the original author)

**Comment by heychr** [comment]
Hello M. Andras, thank you I will try this, currently I am trying to
implement a RL model for multi-path routing, can you please let me know how
the output of the agents (weights of the links) can be applied on the
simulation? What is the parameter that can be responsible to apply the
action received from the agents?

I would really appreciate your insights on this because I am using Omnet++
in my PhD,

thank you,
Christina
